src/hdmf/data_utils.py

```diff
@@ -125,6 +127,272 @@ def maxshape(self):
+        ), "Only one of 'chunk_mb' or 'chunk_shape' can be specified!"
+
+        self._maxshape = self._get_maxshape()
+        self._dtype = self._get_dtype()
+        if chunk_shape is None:
+            self.chunk_shape = self._get_default_chunk_shape(chunk_mb=chunk_mb)
+        else:
+            self.chunk_shape = chunk_shape
+        if buffer_shape is None:
+            self.buffer_shape = self._get_default_buffer_shape(buffer_gb=buffer_gb)
+        else:
+            self.buffer_shape = buffer_shape
+            buffer_gb = np.prod(self.buffer_shape) * np.dtype(self._dtype).itemsize / 1e9
+
+        array_chunk_shape = np.array(self.chunk_shape)
+        array_buffer_shape = np.array(self.buffer_shape)
+        array_maxshape = np.array(self.maxshape)
+        assert all(array_buffer_shape > 0), f"Some dimensions of buffer_shape ({self.buffer_shape}) are less than zero!"
+        assert all(
+            array_buffer_shape <= array_maxshape
+        ), f"Some dimensions of buffer_shape ({self.buffer_shape}) exceed the data dimensions ({self.maxshape})!"
+        assert all(
+            array_chunk_shape <= array_buffer_shape
+        ), f"Some dimensions of chunk_shape ({self.chunk_shape}) exceed the manual buffer shape ({self.buffer_shape})!"
+        assert all((array_buffer_shape % array_chunk_shape == 0)[array_buffer_shape != array_maxshape]), (
+            f"Some dimensions of chunk_shape ({self.chunk_shape}) do not "
+            f"evenly divide the buffer shape ({self.buffer_shape})!"
+        )
+
+        self.num_buffers = np.prod(np.ceil(array_maxshape / array_buffer_shape))
+        self.buffer_selection_generator = (
+            tuple([slice(lower_bound, upper_bound) for lower_bound, upper_bound in zip(lower_bounds, upper_bounds)])
+            for lower_bounds, upper_bounds in zip(
+                product(
+                    *[
+                        range(0, max_shape_axis, buffer_shape_axis)
+                        for max_shape_axis, buffer_shape_axis in zip(self.maxshape, self.buffer_shape)
+                    ]
+                ),
+                product(
+                    *[
+                        chain(range(buffer_shape_axis, max_shape_axis, buffer_shape_axis), [max_shape_axis])
+                        for max_shape_axis, buffer_shape_axis in zip(self.maxshape, self.buffer_shape)
+                    ]
+                ),
+            )
+        )
+
+        if self.display_progress:
+            if self.progress_bar_options is None:
+                self.progress_bar_options = dict()
+
+            try:
+                from tqdm import tqdm
+
+                if "total" in self.progress_bar_options:
+                    warn("Option 'total' in 'progress_bar_options' is not allowed to be over-written! Ignoring.")
+                    self.progress_bar_options.pop("total")
+                self.progress_bar = tqdm(total=self.num_buffers, **self.progress_bar_options)
+            except ImportError:
```

Comment:
Since it is not so easy to test both when tqdm is importable and when it is not, I recommend adding this comment which removes this section from consideration when computing code coverage.
```suggestion
            except ImportError:  # pragma: no cover
```